File: src/fastmath/DoubleColMatrix.java
package fastmath;

import static java.lang.Math.abs;
import static java.lang.Math.floorDiv;
import static java.lang.System.getProperties;

import java.nio.ByteBuffer;
import java.util.HashMap;
import java.util.List;
import java.util.function.BiFunction;

import com.sleepycat.persist.model.Persistent;
import com.sun.jna.Native;
import com.sun.jna.Pointer;

import fastmath.exceptions.NotInvertableException;
import fastmath.matfile.MiDouble;

@Persistent
public class DoubleColMatrix extends DoubleMatrix
{

  private int columnCapacity;

  private final int baseOffset;

  /**
   * When this{@link #appendColumn(Vector)} is called and there isn't enough
   * capacity in the underlying {@link DirectByteBuffer} then a new
   * {@link DirectByteBuffer} is created with a capacity of (1 + this
   * {@link #incrementalCapacityExpasionFactor} ) * this{@link #getColCount()}
   * and the existing memory contents copied over and the previous buffer
   * returned to the heap
   */
  private double incrementalCapacityExpasionFactor = 1.75;

  public DoubleColMatrix()
  {
    this( 0, 0 );
    columnCapacity = 0;
  }

  public DoubleColMatrix(int m, int n)
  {
    super( m, n );
    columnCapacity = n;
    baseOffset = 0;
  }

  public DoubleColMatrix(int m, int n, BiFunction<Integer, Integer, Double> x)
  {
    super( m, n, x );
    columnCapacity = n;
    baseOffset = 0;
  }

  public DoubleColMatrix(List<Pair<Double, Double>> twoColMatrix)
  {
    this( twoColMatrix.size(), 2 );
    int i = 0;
    for ( Pair<Double, Double> pair : twoColMatrix )
    {
      set( i, 0, pair.left );
      set( i, 1, pair.right );
      i++;
    }
  }

  public DoubleColMatrix(int m, int n, String name)
  {
    this( m, n );
    columnCapacity = n;
    setName( name );
  }

  public DoubleColMatrix(ByteBuffer buffer, int m, int n)
  {
    super( buffer, m, n );
    columnCapacity = n;
    baseOffset = 0;
  }

  public DoubleColMatrix(double[][] arr)
  {
    this( arr.length, arr[0].length );

    // TODO: optimize
    for ( int i = 0; i < numRows; i++ )
    {
      for ( int j = 0; j < numCols; j++ )
      {
        set( i, j, arr[i][j] );
      }
    }
  }

  /**
   * Constructs a new matrix, copied from X
   */
  public DoubleColMatrix(AbstractMatrix x)
  {
    super( x.getRowCount(), x.getColCount() );
    baseOffset = 0;
    columnCapacity = getColCount();
    setName( x.getName() );
    for ( int i = 0; i < numCols; i++ )
    {
      Vector src = x.col( i );
      Vector dst = col( i );
      dst.assign( src );
    }
  }

  public DoubleColMatrix(ByteBuffer buffer, int baseOffset, int numRows, int numCols)
  {
    super( buffer, numRows, numCols );
    columnCapacity = numCols;
    this.baseOffset = baseOffset;
  }

  public DoubleColMatrix(String string, double[][] ds)
  {
    this( ds );
    setName( string );
  }

  /**
   * Create a copy of this matrix
   */
  @SuppressWarnings("unchecked")
  @Override
  public DoubleColMatrix copy( boolean reuseBuffer )
  {
    return reuseBuffer ? new DoubleColMatrix( buffer, getBaseOffset(), numRows, numCols ) : new DoubleColMatrix( this );
  }

  /**
   * Create a copy of this matrix
   */
  @Override
  public DoubleColMatrix copy( MatrixContainer container )
  {
    return container.getMatrix( numRows, numCols ).assign( this );
  }

  /**
   * Return offset in BYTES
   * 
   */
  @Override
  public int getOffset( int i, int j )
  {
    assert i >= 0 && i <= numRows : "Row=" + i + ", numRows=" + numRows;
    assert j >= 0 && j <= numCols : "Column=" + j + ", numColumns=" + numCols;

    return getBaseOffset() + ( i * MiDouble.BYTES + ( getColIncrement() * j * MiDouble.BYTES ) );
  }

  public int getOffsetRow( int intOffset )
  {
    return floorDiv( intOffset, getColIncrement() );
  }

  public int getOffsetCol( int intOffset )
  {
    return intOffset % getColIncrement();

  }

  /**
   * Offset (in DOUBLES) needed to get to next column
   */
  @Override
  public int getColIncrement()
  {
    return numRows;
  }

  /**
   * Offset in doubles needed to get to next row
   */
  @Override
  public int getRowIncrement()
  {
    return 1;
  }

  /**
   * 
   * @return the newly appended {@link Vector}
   */
  public Vector appendColumn()
  {
    if ( numCols == columnCapacity )
    {
      columnCapacity += getNewColumnsIncrement();
      int prevSize = numRows * numCols;
      int nextSize = numRows * columnCapacity;
      resizeBuffer( prevSize, nextSize );
    }

    numCols++;
    return (Vector) col( numCols - 1 );
  }

  /**
   * Compacts the internal buffer to the minimum size required
   */
  public void trimToSize()
  {
    int prevColCapacity = columnCapacity;
    columnCapacity = numCols;

    resizeBuffer( prevColCapacity * numCols, numRows * numCols );
  }

  private int getNewColumnsIncrement()
  {
    return Math.max( 1, (int) ( numCols * ( incrementalCapacityExpasionFactor ) ) );
  }

  /**
   * Divide each element in this matrix by a constant scalar
   * 
   * @param x
   * 
   * @return this
   */
  @Override
  public DoubleColMatrix divide( double x )
  {
    for ( Vector col : cols() )
    {
      col.divide( x );
    }

    return this;
  }

  /**
   * Construct an identity matrix
   * 
   * @param i
   *          size of matrix
   * @param v
   *          value to put on the diagonal
   * 
   * @return new DenseDoubleMatrix
   */
  public static AbstractMatrix eye( int i, double v )
  {
    DoubleMatrix x = new DoubleColMatrix( i, i );
    x.diag().assign( v );
    return x;
  }

  /**
   * Construct an identity matrix
   * 
   * @param i
   *          size of matrix
   * 
   * @return new DenseDoubleMatrix
   */
  public static AbstractMatrix eye( int i )
  {
    return eye( i, 1.0 );
  }

  @Override
  public int getMainIncrement()
  {
    return getColIncrement();
  }

  public int getBaseOffset()
  {
    return baseOffset;
  }

  public Vector appendColumn( String name )
  {
    Vector newColumn = appendColumn();
    newColumn.setName( name );
    columnMap.put( name, newColumn );
    return newColumn;
  }

  public Vector getColumn( String name )
  {
    return columnMap.get( name );
  }

  private HashMap<String, Vector> columnMap = new HashMap<>();

  /**
   * 
   * @param i
   * @param expand
   *          if true, and the matrix isnt large enough, it resizes to
   *          accommodate the requested column
   * @return
   */
  public Vector col( int i, boolean expand )
  {
    if ( !expand )
    {
      return col( i );
    }
    else
    {
      while (i >= getColCount())
      {
        appendColumn();
      }

      return col( i );
    }

  }

  public Vector row( int i, boolean resize )
  {
    if ( resize && i >= getRowCount() )
    {
      resize( i + 1, getColCount() );
    }
    return row( i );
  }

  public DoubleColMatrix invert() throws NotInvertableException
  {
    IntVector pivots = new IntVector( getRowCount() );
    Pointer rowCount = newIntParam( getRowCount() );
    Pointer colCount = newIntParam( getColCount() );
    IntVector output = new IntVector( 1 );

    Pointer A = Native.getDirectBufferPointer( buffer );
    Pointer outBuffer = Native.getDirectBufferPointer( output.getBuffer() );
    Pointer pivotBuffer = Native.getDirectBufferPointer( pivots.getBuffer() );
    Lapack.instance.dgetrf_( rowCount, colCount, A, rowCount, pivotBuffer, outBuffer );

    if ( output.elementAt( 0 ) > 0 )
    {
      throw new NotInvertableException( output.elementAt( 0 ) );
    }
    else if ( output.elementAt( 0 ) < 0 )
    {
      throw new IllegalArgumentException( "the " + -output.elementAt( 0 ) + "-th argument had an illegal value" );
    }

    Vector workspace = new Vector( 1024 );
    Pointer lwork = newIntParam( workspace.size );
    Lapack.instance.dgetri_( rowCount, A, rowCount, pivotBuffer, workspace.getPointer(), lwork, outBuffer );
    if ( output.elementAt( 0 ) > 0 )
    {
      throw new NotInvertableException( output.elementAt( 0 ) );
    }
    else if ( output.elementAt( 0 ) < 0 )
    {
      throw new IllegalArgumentException( "the " + -output.elementAt( 0 ) + "-th argument had an illegal value" );
    }

    return this;
  }

  private Pointer newIntParam( int rowCount )
  {
    IntVector ib = new IntVector( 1 );
    ib.setElementAt( 0, rowCount );
    return Native.getDirectBufferPointer( ib.getBuffer() );
  }

  public double supNorm()
  {
    return rows().stream( false ).mapToDouble( row -> row.stream().map( val -> abs( val ) ).sum() ).max().getAsDouble();
  }
}

File: fastmath.net/fastmath/DoubleColMatrix.cs
using Sharpen;

namespace fastmath
{
	public class DoubleColMatrix : fastmath.DoubleMatrix
	{
		private int columnCapacity;

		private readonly int baseOffset;

		/// <summary>
		/// When this
		/// <see cref="appendColumn(string)"/>
		/// is called and there isn't enough
		/// capacity in the underlying
		/// <see cref="DirectByteBuffer"/>
		/// then a new
		/// <see cref="DirectByteBuffer"/>
		/// is created with a capacity of (1 + this
		/// <see cref="incrementalCapacityExpasionFactor"/>
		/// ) * this
		/// <see cref="AbstractMatrix.getColCount()"/>
		/// and
		/// the existing memory contents copied over and the previous buffer returned to
		/// the heap
		/// </summary>
		private double incrementalCapacityExpasionFactor = 1.75;

		public DoubleColMatrix()
			: this(0, 0)
		{
			columnCapacity = 0;
		}

		public DoubleColMatrix(int m, int n)
			: base(m, n)
		{
			columnCapacity = n;
			baseOffset = 0;
		}

		public DoubleColMatrix(int m, int n, java.util.function.BiFunction<int, int, double
			> x)
			: base(m, n, x)
		{
			columnCapacity = n;
			baseOffset = 0;
		}

		public DoubleColMatrix(System.Collections.Generic.IList<fastmath.Pair<double, double
			>> twoColMatrix)
			: this(twoColMatrix.Count, 2)
		{
			int i = 0;
			foreach (fastmath.Pair<double, double> pair in twoColMatrix)
			{
				set(i, 0, pair.left);
				set(i, 1, pair.right);
				i++;
			}
		}

		public DoubleColMatrix(int m, int n, string name)
			: this(m, n)
		{
			columnCapacity = n;
			setName(name);
		}

		public DoubleColMatrix(java.nio.ByteBuffer buffer, int m, int n)
			: base(buffer, m, n)
		{
			columnCapacity = n;
			baseOffset = 0;
		}

		public DoubleColMatrix(double[][] arr)
			: this(arr.Length, arr[0].Length)
		{
			// TODO: optimize
			for (int i = 0; i < numRows; i++)
			{
				for (int j = 0; j < numCols; j++)
				{
					set(i, j, arr[i][j]);
				}
			}
		}

		/// <summary>Constructs a new matrix, copied from X</summary>
		public DoubleColMatrix(fastmath.AbstractMatrix x)
			: base(x.getRowCount(), x.getColCount())
		{
			baseOffset = 0;
			columnCapacity = getColCount();
			setName(x.getName());
			for (int i = 0; i < numCols; i++)
			{
				fastmath.Vector src = x.col(i);
				fastmath.Vector dst = col(i);
				dst.assign(src);
			}
		}

		public DoubleColMatrix(java.nio.ByteBuffer buffer, int baseOffset, int numRows, int
			 numCols)
			: base(buffer, numRows, numCols)
		{
			columnCapacity = numCols;
			this.baseOffset = baseOffset;
		}

		public DoubleColMatrix(string @string, double[][] ds)
			: this(ds)
		{
			setName(@string);
		}

		/// <summary>Create a copy of this matrix</summary>
		public override M copy(bool reuseBuffer)
		{
			return reuseBuffer ? new fastmath.DoubleColMatrix(buffer, getBaseOffset(), numRows
				, numCols) : new fastmath.DoubleColMatrix(this);
		}

		/// <summary>Return offset in BYTES</summary>
		public override int getOffset(int i, int j)
		{
			System.Diagnostics.Debug.Assert(i >= 0 && i <= numRows, "Row=" + i + ", numRows="
				 + numRows);
			System.Diagnostics.Debug.Assert(j >= 0 && j <= numCols, "Column=" + j + ", numColumns="
				 + numCols);
			return getBaseOffset() + (i * fastmath.matfile.MiDouble.BYTES + (getColIncrement(
				) * j * fastmath.matfile.MiDouble.BYTES));
		}

		public virtual int getOffsetRow(int intOffset)
		{
			return System.Math.floorDiv(intOffset, getColIncrement());
		}

		public virtual int getOffsetCol(int intOffset)
		{
			return intOffset % getColIncrement();
		}

		/// <summary>Offset (in DOUBLES) needed to get to next column</summary>
		public override int getColIncrement()
		{
			return numRows;
		}

		/// <summary>Offset in doubles needed to get to next row</summary>
		public override int getRowIncrement()
		{
			return 1;
		}

		/// <returns>
		/// the newly appended
		/// <see cref="Vector"/>
		/// </returns>
		public virtual fastmath.Vector appendColumn()
		{
			if (numCols == columnCapacity)
			{
				columnCapacity += getNewColumnsIncrement();
				int prevSize = numRows * numCols;
				int nextSize = numRows * columnCapacity;
				resizeBuffer(prevSize, nextSize);
			}
			numCols++;
			return (fastmath.Vector)col(numCols - 1);
		}

		/// <summary>Compacts the internal buffer to the minimum size required</summary>
		public virtual void trimToSize()
		{
			int prevColCapacity = columnCapacity;
			columnCapacity = numCols;
			resizeBuffer(prevColCapacity * numCols, numRows * numCols);
		}

		private int getNewColumnsIncrement()
		{
			return System.Math.max(1, (int)(numCols * (incrementalCapacityExpasionFactor)));
		}

		/// <summary>Divide each element in this matrix by a constant scalar</summary>
		/// <param name="x"/>
		/// <returns>this</returns>
		public override X divide(double x)
		{
			foreach (fastmath.Vector col in cols())
			{
				col.divide(x);
			}
			return this;
		}

		/// <summary>Construct an identity matrix</summary>
		/// <param name="i">size of matrix</param>
		/// <param name="v">value to put on the diagonal</param>
		/// <returns>new DenseDoubleMatrix</returns>
		public static fastmath.AbstractMatrix eye(int i, double v)
		{
			fastmath.DoubleMatrix x = new fastmath.DoubleColMatrix(i, i);
			x.diag().assign(v);
			return x;
		}

		/// <summary>Construct an identity matrix</summary>
		/// <param name="i">size of matrix</param>
		/// <returns>new DenseDoubleMatrix</returns>
		public static fastmath.AbstractMatrix eye(int i)
		{
			return eye(i, 1.0);
		}

		public override int getMainIncrement()
		{
			return getColIncrement();
		}

		public virtual int getBaseOffset()
		{
			return baseOffset;
		}

		public virtual fastmath.Vector appendColumn(string name)
		{
			fastmath.Vector newColumn = appendColumn();
			newColumn.setName(name);
			columnMap[name] = newColumn;
			return newColumn;
		}

		public virtual fastmath.Vector getColumn(string name)
		{
			return columnMap[name];
		}

		private System.Collections.Generic.Dictionary<string, fastmath.Vector> columnMap = 
			new System.Collections.Generic.Dictionary<string, fastmath.Vector>();

		/// <param name="i"/>
		/// <param name="expand">
		/// if true, and the matrix isnt large enough, it resizes to accommodate
		/// the requested column
		/// </param>
		/// <returns/>
		public virtual fastmath.Vector col(int i, bool expand)
		{
			if (!expand)
			{
				return col(i);
			}
			else
			{
				while (i >= getColCount())
				{
					appendColumn();
				}
				return col(i);
			}
		}

		public virtual fastmath.Vector row(int i, bool resize)
		{
			if (resize && i >= getRowCount())
			{
				resize(i + 1, getColCount());
			}
			return row(i);
		}

		/// <exception cref="fastmath.exceptions.NotInvertableException"/>
		public virtual fastmath.DoubleColMatrix invert()
		{
			fastmath.IntVector pivots = new fastmath.IntVector(getRowCount());
			com.sun.jna.Pointer rowCount = newIntParam(getRowCount());
			com.sun.jna.Pointer colCount = newIntParam(getColCount());
			fastmath.IntVector output = new fastmath.IntVector(1);
			com.sun.jna.Pointer outBuffer = com.sun.jna.Native.getDirectBufferPointer(output.
				getBuffer());
			com.sun.jna.Pointer pivotBuffer = com.sun.jna.Native.getDirectBufferPointer(pivots
				.getBuffer());
			fastmath.LAPACK.instance.dgetrf_(rowCount, colCount, getPointer(), rowCount, pivotBuffer
				, outBuffer);
			if (output.elementAt(0) > 0)
			{
				throw new fastmath.exceptions.NotInvertableException(output.elementAt(0));
			}
			else
			{
				if (output.elementAt(0) < 0)
				{
					throw new System.ArgumentException("the " + -output.elementAt(0) + "-th argument had an illegal value"
						);
				}
			}
			fastmath.Vector workspace = new fastmath.Vector(1024);
			com.sun.jna.Pointer lwork = newIntParam(workspace.size);
			fastmath.LAPACK.instance.dgetri_(rowCount, getPointer(), rowCount, pivotBuffer, workspace
				.getPointer(), lwork, outBuffer);
			if (output.elementAt(0) > 0)
			{
				throw new fastmath.exceptions.NotInvertableException(output.elementAt(0));
			}
			else
			{
				if (output.elementAt(0) < 0)
				{
					throw new System.ArgumentException("the " + -output.elementAt(0) + "-th argument had an illegal value"
						);
				}
			}
			return this;
		}

		private com.sun.jna.Pointer newIntParam(int rowCount)
		{
			fastmath.IntVector ib = new fastmath.IntVector(1);
			ib.setElementAt(0, rowCount);
			return com.sun.jna.Native.getDirectBufferPointer(ib.getBuffer());
		}

		public virtual double supNorm()
		{
			return rows().stream(false).mapToDouble(null).max().getAsDouble();
		}

		public virtual fastmath.DoubleColMatrix assign(double[][] ds)
		{
			for (int m = 0; m < numRows; m++)
			{
				for (int n = 0; n < numCols; n++)
				{
					set(m, n, ds[m][n]);
				}
			}
			return this;
		}
	}
}
